core/lib/zksync_core/src/eth_watch/event_processors/set_chain_id.rs

```diff
@@ -0,0 +1,73 @@
+use zksync_contracts::SET_CHAIN_ID_EVENT;
+use zksync_dal::StorageProcessor;
+use zksync_types::{protocol_version::decode_set_chain_id_event, web3::types::Log, Address, H256};
+
+use crate::eth_watch::{
+    client::{Error, EthClient},
+    event_processors::EventProcessor,
+    metrics::{PollStage, METRICS},
+};
+
+/// Responsible for saving `setChainId` upgrade transactions to the database.
+#[derive(Debug)]
+pub struct SetChainIDEventProcessor {
+    /// Address of the `DiamondProxy` contract of the chain that the `SetChainId` event is for.
+    diamond_proxy_address: Address,
+    /// Signature of the `SetChainIdUpgrade` event.
+    /// The event is emitted by the `StateTransitionManager` contract.
+    set_chain_id_signature: H256,
+}
+
+impl SetChainIDEventProcessor {
+    pub fn new(diamond_proxy_address: Address) -> Self {
+        Self {
+            diamond_proxy_address,
+            set_chain_id_signature: SET_CHAIN_ID_EVENT.signature(),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl EventProcessor for SetChainIDEventProcessor {
+    async fn process_events(
+        &mut self,
+        storage: &mut StorageProcessor<'_>,
+        _client: &dyn EthClient,
+        events: Vec<Log>,
+    ) -> Result<usize, Error> {
+        // SetChainId does not go through the governance contract, so we need to parse it separately.
+        let upgrades = events
+            .into_iter()
+            .filter(|log| {
+                log.topics[0] == self.set_chain_id_signature
+                    && log.topics[1] == self.diamond_proxy_address.into()
+            })
+            .map(|event| {
+                decode_set_chain_id_event(event)
+                    .map_err(|err| Error::LogParse(format!("{:?}", err)))
```

Comment:
Off-topic: I feel that `LogParse` wraps a wrong type (should be some error, maybe `L1TxParseError`), but that has little to do with the PR 🙃